generic-contracts/src/main/java/com/scalar/dl/genericcontracts/table/v1_0_0/Insert.java

```diff
@@ -0,0 +1,115 @@
+    }
+
+    // Check the key existence and type in the argument
+    String key = tableAsset.get().data().get(Constants.TABLE_KEY).textValue();
+    String keyType = tableAsset.get().data().get(Constants.TABLE_KEY_TYPE).textValue();
+    JsonNode values = arguments.get(Constants.RECORD_VALUES);
+    if (!values.has(key)) {
+      throw new ContractContextException(Constants.RECORD_KEY_NOT_EXIST);
+    }
+    if (!keyType.toUpperCase().equals(values.get(key).getNodeType().name())) {
+      throw new ContractContextException(Constants.INVALID_KEY_TYPE);
+    }
+
+    // Check the record existence
+    String recordAssetId = getAssetIdForRecord(tableName, key, values.get(key).asText());
+    Optional<Asset<JsonNode>> recordAsset = ledger.get(recordAssetId);
+    if (recordAsset.isPresent()) {
+      throw new ContractContextException(Constants.RECORD_ALREADY_EXISTS);
+    }
+
+    // Put the asset records
+    putIndexAssets(
+        ledger,
+        tableName,
+        key,
+        arguments.get(Constants.RECORD_VALUES),
+        tableAsset.get().data().get(Constants.TABLE_INDEXES));
+    ledger.put(recordAssetId, arguments.get(Constants.RECORD_VALUES));
+
+    return null;
+  }
+
+  private void putIndexAssets(
+      Ledger<JsonNode> ledger, String tableName, String key, JsonNode values, JsonNode indexes) {
+    for (JsonNode index : indexes) {
+      String indexKey = index.get(Constants.INDEX_KEY).asText();
+      String indexKeyType = index.get(Constants.INDEX_KEY_TYPE).asText();
+      String assetId;
+
+      if (values.has(indexKey) && !values.get(indexKey).isNull()) {
+        if (!indexKeyType.toUpperCase().equals(values.get(indexKey).getNodeType().name())) {
+          throw new ContractContextException(Constants.INVALID_INDEX_KEY_TYPE);
+        }
+        assetId = getAssetIdForIndex(tableName, indexKey, values.get(indexKey).asText());
+      } else {
+        assetId = getAssetIdForNullIndex(tableName, indexKey);
+      }
+
+      ObjectNode node = getObjectMapper().createObjectNode();
+      node.set(key, values.get(key));
+      node.put(Constants.ASSET_AGE, 0);
+
+      ledger.put(assetId, node);
+    }
+  }
+
+  private String getAssetIdForTable(String tableName) {
+    return Constants.PREFIX_TABLE + tableName;
+  }
+
+  private String getAssetIdForRecord(String tableName, String keyColumnName, String key) {
+    return Constants.PREFIX_RECORD
+        + tableName
+        + Constants.ASSET_ID_SEPARATOR
+        + keyColumnName
+        + Constants.ASSET_ID_SEPARATOR
+        + key;
+  }
+
+  private String getAssetIdForIndex(String tableName, String indexKey, String value) {
+    return Constants.PREFIX_INDEX
+        + tableName
+        + Constants.ASSET_ID_SEPARATOR
+        + indexKey
+        + Constants.ASSET_ID_SEPARATOR
+        + value;
+  }
+
+  private String getAssetIdForNullIndex(String tableName, String indexKey) {
+    return Constants.PREFIX_INDEX + tableName + Constants.ASSET_ID_SEPARATOR + indexKey;
```

Comment:
I'm wondering if this `NullIndex` is for.
(I might miss something, but it's not discussed in the design doc?)
Anyway, it looks like it's for scanning an index?

Reply:
@feeblefakie Thanks for the question! Yes, it’s for the index scan. We are going to accept a condition where an index value is null. Although it was described in # 2 of [this section](https://docs.google.com/document/d/1lrpOKbVedrIVcee5kkiR2MItmNlEEDO_CLw7FfbS4uo/edit?pli=1&tab=t.0#heading=h.pdnpgt8r1uzd), I’ve added related descriptions in both the Insert and Select contracts sections for clarification.

Reply:
Ah, I kind of misunderstood it.
I got it. Thank you!